src/SignalR/clients/csharp/Client.SourceGenerator/src/HubProxyGenerator.cs

```diff
@@ -0,0 +1,256 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.AspNetCore.SignalR.Client.SourceGenerator
+{
+    [Generator]
+    internal class HubProxyGenerator : ISourceGenerator
+    {
+        public void Initialize(GeneratorInitializationContext context)
+        {
+            context.RegisterForSyntaxNotifications(() => new SyntaxReceiver());
+        }
+
+        public void Execute(GeneratorExecutionContext context)
+        {
+            var getProxyBody = new StringBuilder(@"// Generated by Microsoft.AspNetCore.Client.SourceGenerator
+using Microsoft.AspNetCore.SignalR.Client;
+namespace Microsoft.AspNetCore.SignalR.Client
+{
+    public static partial class HubConnectionExtensionsGenerated
+    {
+        public static THub GetProxy<THub>(this IHubConnection conn)
+        {");
+            var getProxyFooter = @"
+            throw new System.ArgumentException(nameof(THub));
+        }
+    }
+}";
+
+            if (context.SyntaxReceiver is not SyntaxReceiver receiver)
+            {
+                // nothing to do yet
+                getProxyBody.Append(getProxyFooter);
+                context.AddSource("GeneratedBy" + nameof(HubProxyGenerator),
+                    SourceText.From(getProxyBody.ToString(), Encoding.UTF8));
+                return;
+            }
+
+            var hubSymbols = new Dictionary<string, (ITypeSymbol, MemberAccessExpressionSyntax)>();
+            var iHubConnectionType =
+                context.Compilation.GetTypeByMetadataName("Microsoft.AspNetCore.SignalR.Client.IHubConnection");
+
+            // Go thru candidates and filter
+            foreach(var (memberAccess, proxyType) in receiver.Candidates)
+            {
+                var expressionModel = context.Compilation.GetSemanticModel(memberAccess.Expression.SyntaxTree);
+                var typeInfo = expressionModel.GetTypeInfo(memberAccess.Expression);
+
+                // Filter based on receiver symbol
+                if(!SymbolEqualityComparer.Default.Equals(typeInfo.Type, iHubConnectionType) &&
+                   (typeInfo.Type.AllInterfaces.IsDefaultOrEmpty ||
+                    !typeInfo.Type.AllInterfaces.Any(x => SymbolEqualityComparer.Default.Equals(x, iHubConnectionType))))
```

Comment:
Incorporated @davidfowl's suggestion for checking the receiver type of `GetProxy<THub>` extension method to get a stronger guarantee than just checking the name. I'm still looking for edge cases here but seems solid so far.